File: ruby/567.permutation-in-string.rb
# @param {String} s1
# @param {String} s2
# @return {Boolean}
def check_inclusion(s1, s2)
  hash = s1.chars.tally

  left = 0
  right = left

  while right < s2.length
    return true if hash.values.sum.zero?

    char = s2[right]
    if hash[char].to_i > 0
      hash[char] -= 1
      right += 1
    else
      hash[s2[left]] += 1 if hash.key?(s2[left])

      right += 1 if left == right
      left += 1
    end
  end
  return true if hash.values.sum.zero?

  false
end

# left < rightであるという保証が自分のコードにはない
# 存在しなかった場合、hashに何かしらの処理をしてkら left += 1するはず？

File: rust/src/567.permutation-in-string.rs
impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let mut freq = [0; 26];
        let s1_bytes = s1.as_bytes();
        let s2_bytes = s2.as_bytes();

        for &byte in s1_bytes {
            freq[(byte - b'a') as usize] += 1;
        }

        let mut left = 0;
        let mut right = 0;

        while right < s2_bytes.len() {
            let idx = (s2_bytes[right] - b'a') as usize;

            if freq[idx] > 0 {
                freq[idx] -= 1;
                right += 1;
            } else if left == right {
                left += 1;
                right += 1;
            } else {
                freq[(s2_bytes[left] - b'a') as usize] += 1;
                left += 1;
            }

            if right - left == s1_bytes.len() {
                return true;
            }
        }

        false
    }
}
